### ISSUE TITLE
Upgrade Routine for Customer Purchase IDs

### ISSUE THREAD

**Issue by cklosowski** [easydigitaldownloads/Easy-Digital-Downloads#3107]
The initial upgrade routine of customers (v2.2) apparently missed the attachment of some purchase IDs to the customers.

As we've now focused on the `EDD_Customer` class in #2548 && #3018, we need to make sure that for each Customer, we search for payment records not attached.

We can do this by:
Getting a list of purchase IDs currently attached to the customer ID
Getting a list of all purchase IDs associated to the email address of the customer
If the customer is attached to a user id, searching for payments attached to that user id.

From there, we can get all the unique IDs and we should be covered.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by cklosowski** [comment]
@pippinsplugins Should we make sure the purcahse_count and purchase_value match up too? Trying to come up with a good way to do this without wrecking the DB with updates.

**Comment by pippinsplugins** [comment]
I don't think so. As long as they get incremented at the same time (for earnings and sales), they should match up.

**Comment by cklosowski** [comment]
ok so just attaching payment IDs to the customer, I can do that.

**Comment by pippinsplugins** [comment]
To test this, I'm going to manually remove some payments from the `payment_ids` column for some customers, then run the routine and see if they get added properly.

**Comment by pippinsplugins** [comment]
I think we should go ahead and add `Doing step X of Y` to this. For EDD 2.2, we had a couple of tickets where people thought the routine had stalled because it just kept going. In reality, there was just a lot of steps to complete. Showing the step number will give people reassurance it's working as intended.

**Comment by cklosowski** [comment]
@pippinsplugins yeah that sounds fine. I'll add in the steps here.

**Comment by pippinsplugins** [comment]
We need to make sure that customer purchase stats are reset to 0 if no payments are found.

**Comment by pippinsplugins** [comment]
There's a bug in this that has caused really, really wrong stats for some customers:
![screen shot 2015-02-17 at 1 47 09 pm](https://cloud.githubusercontent.com/assets/1034109/6235911/95e3e65c-b6ab-11e4-99d0-80bd8eb5f1fc.png)

**Comment by cklosowski** [comment]
@pippinsplugins I think the last thing to do here is use this as a reminder to go through all of `release/2.3` and find places we use `$customer->user_id` and verify our checks aren't just looking for `empty()`, then we can close it back out.

**Comment by pippinsplugins** [comment]
I think I've got them all.

**Comment by pippinsplugins** [comment]
I'm not sure if this is due to the upgrade or a bug in the purchase creation process, but I just got this report from a customer on pippinsplugins:

>>>I  just purchased an unlimited license for Easy Content Types, and when I returned to the Account page, none of my other plugin purchases are listed. The Plugin Purchase History list only has my last purchase.

I looked up their purchase history and all payments (5 of them) show as being connected to the customer when viewing the `View Order Details` screen, but the customer details screen shows just one (the same one he sees in his account):

![screen shot 2015-02-25 at 10 21 03 am](https://cloud.githubusercontent.com/assets/1034109/6374673/0849d382-bcd8-11e4-9072-b35b27de4398.png)

I looked up his customer record and sure enough, there is only one payment ID in the `payment_ids` column.

A few things:

1. We need to test this a few more times to see if we can find what's causing the disconnect. My guess is that there is a bug in the `attach_payment()` method.
2. I think we should adjust the update payment action to always check to see if the payment ID being updated exists in the `payment_ids` column of the connected customer, and if it doesn't, add it. That will allow site admins to fix cases like this by simply clicking `Update Payment` on `View Order Details`.

**Comment by pippinsplugins** [comment]
Ok the `attach_payment()` method seems to work fine. I just ran the following for each missing payment and it worked fine:
```php
$customer = new EDD_Customer( $email );
var_dump( $customer->attach_payment( 43437, false ) );
```

**Comment by cklosowski** [comment]
Ok Tracked it down to the `edd_insert_payment()` function. It was treating new customers and existing customers the same and overwriting their payment_ids column.

We've updated the function to use the `attach_payment` method of `EDD_Customer`, this way all the hooks fire when adding to either a new or existing customer.

**Comment by pippinsplugins** [comment]
Fixed the payment overwriting issue in 82374483265fb6f2b3cffaf5be16b56fce539b53

**Comment by pippinsplugins** [comment]
Took 8:06 minutes to complete this on a live site that had 38 steps.
